karapace/kafka_admin.py

```diff
@@ -0,0 +1,290 @@
+"""
+Copyright (c) 2023 Aiven Ltd
+See LICENSE for details
+"""
+
+from __future__ import annotations
+
+from collections.abc import Iterable
+from concurrent.futures import Future
+from confluent_kafka import TopicPartition
+from confluent_kafka.admin import (
+    AdminClient,
+    BrokerMetadata,
+    ClusterMetadata,
+    ConfigResource,
+    ConfigSource,
+    NewTopic,
+    OffsetSpec,
+    ResourceType,
+    TopicMetadata,
+)
+from confluent_kafka.error import KafkaException
+from kafka.errors import AuthenticationFailedError, for_code, UnknownTopicOrPartitionError
+from karapace.constants import TOPIC_CREATION_TIMEOUT_S
+from typing import Any, Callable, Container, NoReturn, Protocol, TypedDict, TypeVar
+from typing_extensions import Unpack
+
+import logging
+
+LOG = logging.getLogger(__name__)
+
+
+T = TypeVar("T")
+
+
+def single_futmap_result(futmap: dict[Any, Future[T]]) -> T:
+    """Extract the result of a future wrapped in a dict.
+
+    Bulk operations of the `confluent_kafka` library's Kafka clients return results
+    wrapped in a dictionary of futures. Most often we use these bulk operations to
+    operate on a single resource/entity. This function makes sure the dictionary of
+    futures contains a single future and returns its result.
+    """
+    assert len(futmap) == 1
+
+    return next(iter(futmap.values())).result()
+
+
+def raise_from_kafkaexception(exc: KafkaException) -> NoReturn:
+    """Raises a more developer-friendly error from a `KafkaException`.
+
+    The confluent-kafka library's `KafkaException` is a wrapper around its internal
+    `KafkaError`. The resulting, raised exception however is coming from
+    `kafka-python`, due to these exceptions having human-readable names, providing
+    better context for error handling.
```

Comment:
issue: I'm weary about this, it looks like this will tie us into a future where we'll never be able to get rid of kafka-python, forever having to depend on both implementations.

Reply:
I agree, however, I think overhauling the error handling here could be postponed until we are actually at the point when we can remove all usages of the `kafka-python` library. Until then, I think it's simpler to make use of this convenience, without having to adapt clients and wrap error-codes in meaningful exceptions, etc. (as not to have to check error-codes in client code).

Reply:
That sounds fair to split up the effort 👍